### ISSUE TITLE
@RequestMapping without @Controller registered as handler [SPR-17622]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-framework#22154]
**[Eugene Tenkaev](https://jira.spring.io/secure/ViewProfile.jspa?name=hronom)** opened **[SPR-17622](https://jira.spring.io/browse/SPR-17622?redirect=false)** and commented

Following this approach here http://projects.spring.io/spring-cloud/spring-cloud.html#spring-cloud-feign-inheritance

If you add root `@RequestMapping` to the `UserService` it will be registered as handler in Spring MVC application.

Example project to reproduce here https://github.com/Hronom/test-shared-mapping-interface

Related discussions:
* https://github.com/spring-cloud/spring-cloud-netflix/issues/466
* https://stackoverflow.com/questions/29284911/can-a-spring-cloud-feign-client-share-interface-with-an-spring-web-controller

To handle this properly need to avoid registration of controller that has only `@RequestMapping` annotation.

Proposed solution:
Register handler only if it has annotation `@Controller` or `@RestController`.

---

**Affects:** 5.1.3

**Issue Links:**
- #16747 Introduce proxy-based REST client similar to HttpInvokerProxyFactoryBean

1 votes, 3 watchers

**Comment by rstoyanchev** [comment]
Class level `@RequestMapping` is used as a hint, independent of `@Controller`, because `@RequestMapping` can be used on an interface, in which case the controller can be an AOP proxy and the `@Controller` annotation is not accessible to Spring MVC through the proxy.

@jhoeller do you see any options to refine the checks, e.g. if type-level `@RequestMapping` is found without `@Controller` and the bean is a proxy, then introspect further to see if we can find the `@Controller` annotation?

**Comment by remal** [comment]
This issue leads to a lot of different problems that are very hard to debug. Please fix it. I can suggest these solutions:
1. Do not treat classes annotated by `@RequestMapping` as handlers. Only `@Controller` annotation should be taken into consideration.
1. Spring Data has `@NoRepositoryBean`. A similar annotation can be created for request handlers. For example: `@NoRequestHandler`.

I'd suggest implementing the first solution.

**Comment by TannnnnnnnnnnnnnnnK** [comment]
regist handler and regist requestmapping
two different things, should be separate from each other, even they are related

**Comment by glockbender** [comment]
This issue is still relevant. Any progress with that?

**Comment by odrotbohm** [comment]
Copying the description of #25386 here for reference:

**tl;dr** – The current behavior is problematic for folks trying to customize Spring Data REST using custom controllers, too, as it subtley makes controllers using `@RequestMapping` on the type level end up in the wrong handler mapping

> `RequestMappingHandlerMapping.isHandler(…)` not only picks up types annotated with `@Controller` but also ones that are annotated with `@RequestMapping` on the type level. This is problematic in cases in which other `HandlerMapping` instances are registered that might be supposed to handle those controllers.
>
> A prominent example of this is Spring Data REST, which registers a dedicated mapping to expose HTTP resources for Spring Data repositories. Users can selectively override those resources by declaring a controller themselves and just declare a handler method for e.g. the URI of an item resource and an HTTP verb of choice. If that controller now declares an `@RequestMapping` on the type level, the Spring MVC registered one will pick up that class, and not see any other mappings defined for the same URI pattern but exposing support for other HTTP methods potentially available in subsequent `HanderMapping` implementations.
> 
> This is a pretty common error scenario reported by users (see [this StackOverflow](https://stackoverflow.com/questions/62865947/restrepositorycontroller-hide-rest-repository-endpoints/62877864) question for example). It's also pretty hard to explain to users as it involves talking about quite a few implementation details.
>
> Removing the explicit handling of `@RequestMapping` on the type level bears the risk that controller implementations not also being annotated with `@Controller` would not be picked up automatically anymore. I haven't found any Spring MVC related documentation that actually shows an example of code not using the annotations in combination when used at the type level. A fix for that issue would be to also annotate the affected controller type with `@Controller`. I can see this being suboptimal for a release in a minor version but for 6.0 we should at least reevaluate.

**Comment by mothinx** [comment]
Just get this issue with a feign client and spent a lot of hours to locate it. Is someone on that issue ?

**Comment by odrotbohm** [comment]
This just came up in yet another Spring Data REST issue: https://jira.spring.io/browse/DATAREST-1591.
